### ISSUE TITLE
possible memory leak?

### ISSUE THREAD

**Issue by leeoniya** [domvm/domvm#229]
this started appearing recently (perhaps after a Chrome bump), which is a bit concerning:

![image](https://user-images.githubusercontent.com/43234/126829517-cccc20ae-62f5-498d-a427-06d2d3102591.png)

https://krausest.github.io/js-framework-benchmark/current.html

rather odd since nothing has changed in the lib.

@lawrence-dol might have some fond memories of us tackling mem leaks in firefox back in the day. perhaps if you have time, you can take a peek to see if it's real or being misreported or GC is just being lazy?

**Comment by lawrence-dol** [comment]
I happen to have an auto-replay feature for one of my apps, although it uses a grid layout instead of a table. It shows somewhat delayed GC after activity ends; but no leaks -- all metrics return to initial levels. I don't have anything handy that will do repeated manipulations on a table.

I've outline the replay of 1000 iterations (full display updates). I'm actually a little fascinated that it maxed out my 12-core/24-thread machine.

![image](https://user-images.githubusercontent.com/8662551/126840913-e51b5bcb-555c-4df2-a877-41b97c8a78ec.png)

**Comment by leeoniya** [comment]
thanks for checking! :100:

**Comment by lawrence-dol** [comment]
One thing I noted is that Chromium browsers appear to "leak documents". That is each time I refresh the page containing the app, the documents count increases, and it never seems to return to 1. Maybe the testing is faulty in that way? Or maybe DOMVM leaves a reference to the page reachable (though I can't suggest how it could do that).

**Comment by lawrence-dol** [comment]
@leeoniya : Actually, there may be a problem with tables, specifically. 

I have a key-event diagnostic tool (at http://tech.dolhub.com/utility/KeyEvents) and it shows unexpectedly high DOM node counts in the browser's Performance Monitor. It's currently running with the developer build to see if I can figure out what's going on.

Each key press/release should result in the addition of 26 nodes (1 tr and 12 td's), and the DOMVM console log bears this out; but the performance monitor shows the node count increasing by about 50 nodes -- not quite twice that. Each time I press and release a key DOMVM consistently reports adding 26 elements.

***But***... if I filter out keydown events, which reduces the number of rows in the list by half, DOMVM reports a removeChild for each row but the DOM Nodes count in the Performance Monitor doesn't change. It does look like the row elements are retained despite being removed from the view.

[DOMVM 3.4.11]

**Comment by leeoniya** [comment]
yeah, odd. haven't looked at the code, but this is what i get when pressing "1" a lot, then pressing clear, repeat several rounds. then at the end force a GC in devtools and it all gets collected:

![image](https://user-images.githubusercontent.com/43234/127089364-fdabc91a-90a0-4af9-ba90-1bfd758dce65.png)

**Comment by leeoniya** [comment]
there's nothing really specific for tables, but maybe this optimization is not doing enough when the entire body is removed: https://github.com/domvm/domvm/blob/8e0ef3c8a7fae78e3c4a9342b3d8d38c03c0f2a9/src/view/dom.js#L103-L106

you can try removing that optimization and see if it makes a difference. otherwise poking at various functions there like removeChild might yield some insights.

**Comment by lawrence-dol** [comment]
I'm digging into this, though I have customers to attend to, so it will be here and there as time permits (and curiosity gets the better of my discipline).

_Something_ is definitely hinky. I commented out the entire view except the event table. Pressing a single key emits the periodic header row, then the keydown, then the key up. Each row is 12 `td`s, for a total of 13 including the `tr`. DOMVM diags reports exactly the expected DOM operations. But the baseline of (consistently) 52 DOM elements changes +75 to 127 elements (also consistently). Examination of the DOM tree shows exactly and only the expected elements. And forcing GC after doing this has no effect. So the net effect is `75 - 39 = 46` DOM nodes unaccounted for, when nodes were only added not removed.

**Comment by leeoniya** [comment]
thanks!

i'm just trying a simple fill/clear case repeatedly which reproduces the behavior:

```html
<!doctype html>
<html>
    <head>
        <title>mem leak</title>
        <script src="./domvm.nano.iife.js"></script>
    </head>
    <body>
        <script>
            let {
                createView: cv,
                defineView: vw,
                defineElement: el,
            } = domvm;

            let App = (vm, data) => {
                let showAll = false;

                const toggle = e => {
                    showAll = !showAll;
                    vm.redraw();
                };

                return (vm, data) => (
                    el("div", [
                        el("button", {onclick: toggle}, "Toggle"),
                        showAll ? vw(Table, data) : null,
                    ])
                );
            };

            let Table = (vm, data) => {
                return (vm, data) => (
                    el("table", data.map(v =>
                        el("tr", [
                            el("td", v)
                        ])
                    ))
                );
            };

            const data = Array.from({length: 1000}, () => Math.floor(Math.random() * 1000));

            cv(App, data).mount(document.body);
        </script>
    </body>
</html>
```

After a few dozen fill/clear cycles, both both Chrome's and Firefox's heap snapshots show the expected retained objects and heap sizes with a filled and emptied view:

Chrome:

1000 rows:

![image](https://user-images.githubusercontent.com/43234/127264745-cbe7c17a-9ce9-4072-8fcc-f984c1d26f1e.png)

0 rows:

![image](https://user-images.githubusercontent.com/43234/127264884-a606d1a2-87ef-43a7-bc9e-a8ad8d9b0ed3.png)

When watching Chrome's Task Manager (Shift + Esc), i see similar behavior to what's in the recorded performance profile: very lazy memory reclamation, and only under pressure. If i press toggle 40 times and let the tab's mem grow from 28MB -> 46MB, then pause with a filled table, and then click clear, it will reclaim maybe 70% of the difference. then if i force a GC it will reclaim the rest without issues.

so, i don't _think_ there's a real leak or excessive allocation going on here, but for some reason chrome is not as eager to reclaim memory as it is with other libs.

i think it's worth trying to replicate the above code in vanilla js just to get a baseline for what to expect.

**Comment by leeoniya** [comment]
did some more testing with a mostly identical React version. 60s total with clear/fill toggle every 500ms.

the patterns are pretty much the same. node counts grow with vnode heap until the GC decides it's time to reclaim, at which point it reclaims everything. uPlot's overall mem usage is about 40% less than React, which is maybe why it takes longer for the GC pressure to build.

uPlot:

![image](https://user-images.githubusercontent.com/43234/127545190-91f5e580-5c46-4b43-b5c7-2f9fec166680.png)

React:

![image](https://user-images.githubusercontent.com/43234/127545258-d47402e8-c372-4d20-b004-f3f2cf4a4de1.png)


uPlot code:

```html
<!doctype html>
<html>
    <head>
        <title>mem leak</title>
        <script src="./domvm.nano.iife.js"></script>
    </head>
    <body>
        <script>
            let {
                createView: cv,
                defineView: vw,
                defineElement: el,
            } = domvm;

            let App = (vm, data) => {
                let showAll = true;

                const toggle = e => {
                    showAll = !showAll;
                    vm.redraw();
                };

                setTimeout(() => {
                    setInterval(() => {
                        toggle();
                    }, 500);
                }, 2000);

                return (vm, data) => (
                    el("div", [
                        el("button", {onclick: toggle}, "Toggle"),
                        showAll ? vw(Table, data) : null,
                    ])
                );
            };

            let Table = (vm, data) => {
                return (vm, data) => (
                    el("table", data.map(v =>
                        el("tr", [
                            el("td", v)
                        ])
                    ))
                );
            };

            const data = Array.from({length: 1000}, () => Math.floor(Math.random() * 1000));

            cv(App, data).mount(document.body);
        </script>
    </body>
</html>
```

React code:

```jsx
import React from 'react';

export default class TodoApp extends React.Component {
    state = {
        items: Array.from({length: 1000}, () => Math.floor(Math.random() * 1000)),
        show: true,
    };

    constructor(props) {
      super(props)

      setTimeout(() => {
        setInterval(() => {
            this.setState({
                show: !this.state.show
            });
        }, 500);
      }, 2000);
    }

    render() {
        if (this.state.show) {
            return (
                <table>
                {this.state.items.map(item => (
                    <tr>
                        <td>
                            {item}
                        </td>
                    </tr>
                ))}
                </table>
            )
        }

        return <></>;
    }
}
```

**Comment by leeoniya** [comment]
no leaks (see linked issue)
